mapo/agents/mapo/off_mapo_policy.py

```diff
@@ -71,20 +80,34 @@ def _build_critic_loss(policy, batch_tensors):
 
 def _build_actor_loss(policy, batch_tensors):
     obs = batch_tensors[SampleBatch.CUR_OBS]
-    policy_action = policy.model.get_actions(obs)
-    policy_action_value = policy.model.get_q_values(obs, policy_action)
-    policy.loss_stats["qpi_mean"] = tf.reduce_mean(policy_action_value)
-    return -tf.reduce_mean(policy_action_value)
+    gamma = policy.config["gamma"]
+    model = policy.model
+    n_samples = policy.config["branching_factor"]
+
+    policy_action = model.get_actions(obs)
+    next_state_dist = model.next_state_dist(obs, policy_action)
+    sampled_next_state = next_state_dist.sample((n_samples,))
+    next_state_log_prob = tf.reduce_sum(
+        next_state_dist.log_prob(sampled_next_state), axis=-1
+    )
+    next_state_value = tf.stop_gradient(model.get_state_values(sampled_next_state))
+    # later: add reward function gradient before gamma
+    model_aware_policy_loss = tf.reduce_mean(
+        gamma * tf.reduce_mean(next_state_log_prob * next_state_value, axis=0)
+    )
+    return model_aware_policy_loss
 
 
-def build_actor_critic_losses(policy, batch_tensors):
-    """Contruct actor (DPG) and critic (Fitted Q) losses."""
+def build_mapo_losses(policy, batch_tensors):
+    """Contruct actor (MADPG), critic (Fitted Q) and dynamics (MLE) losses."""
     policy.loss_stats = {}
+    policy.dynamics_loss = _build_dynamics_loss(policy, batch_tensors)
```

Comment:
What is the easiest way to have access to the config dict so to choose the appropriate _build_dynamics_loss (between MLE and PG-aware) ?

Reply:
I guess we can just call `policy.config` here, right?

Reply:
That's right. I didn't consider putting it here since the gradient-aware loss wasn't implemented, but I'll do it now.